File: ipaacalib/cpp/test/src/testipaaca.cc
/*
 * This file is part of IPAACA, the
 *  "Incremental Processing Architecture
 *   for Artificial Conversational Agents".  
 *
 * Copyright (c) 2009-2013 Sociable Agents Group
 *                         CITEC, Bielefeld University   
 *
 * http://opensource.cit-ec.de/projects/ipaaca/
 * http://purl.org/net/ipaaca
 *
 * This file may be licensed under the terms of of the
 * GNU Lesser General Public License Version 3 (the ``LGPL''),
 * or (at your option) any later version.
 *
 * Software distributed under the License is distributed
 * on an ``AS IS'' basis, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied. See the LGPL for the specific language
 * governing rights and limitations.
 *
 * You should have received a copy of the LGPL along with this
 * program. If not, go to http://www.gnu.org/licenses/lgpl.html
 * or write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.  
 *
 * The development of this software was supported by the
 * Excellence Cluster EXC 277 Cognitive Interaction Technology.
 * The Excellence Cluster EXC 277 is a grant of the Deutsche
 * Forschungsgemeinschaft (DFG) in the context of the German
 * Excellence Initiative.
 */

#include <ipaaca/ipaaca.h>
#include <typeinfo>

using namespace ipaaca;

const char RECV_CATEGORY[] = "WORD";
const char SEND_CATEGORY[] = "TEXT";

class TextSender {
	protected:
		OutputBuffer::ptr _ob;
		InputBuffer::ptr  _ib;
	public:
		TextSender();
		void outbuffer_handle_iu_event(IUInterface::ptr iu, IUEventType event_type, bool local);
		void inbuffer_handle_iu_event(IUInterface::ptr iu, IUEventType event_type, bool local);
		IUInterface::ptr find_last_iu();
		void publish_text_to_print(const std::string& text, const std::string& parent_iu_uid="");
};

TextSender::TextSender() {
	_ob = OutputBuffer::create("TextSenderOut");
	_ob->register_handler(boost::bind(&TextSender::outbuffer_handle_iu_event, this, _1, _2, _3));
	_ib = InputBuffer::create("TextSenderIn", RECV_CATEGORY);
	_ib->register_handler(boost::bind(&TextSender::inbuffer_handle_iu_event, this, _1, _2, _3));
}

void TextSender::outbuffer_handle_iu_event(IUInterface::ptr iu, IUEventType event_type, bool local)
{
	std::cout << "(own IU event " << iu_event_type_to_str(event_type) << " " << iu->uid() << ")" << std::endl;
	if (event_type == IU_UPDATED) {
		std::set<std::string> parent_uids = iu->get_links("GRIN");
		if (parent_uids.size() > 0) {
			std::string parent_uid = *(parent_uids.begin());
			std::cout << "updating parent ..." << std::endl;
			std::set<std::string> next_uids = iu->get_links("SUCCESSOR");
			if (next_uids.size() > 0) {
				std::string next_uid = *(next_uids.begin());
				IUInterface::ptr next_iu = _ob->get(next_uid);
				std::set<std::string> next_letter_grin_links = next_iu->get_links("GRIN");
				if (next_letter_grin_links.count(parent_uid) == 0) {
					// next letter belongs to new word
					IUInterface::ptr parent_iu = _ib->get(parent_uid);
					parent_iu->payload()["STATE"] = "REALIZED";
				} else {
					IUInterface::ptr parent_iu = _ib->get(parent_uid);
					parent_iu->payload()["STATE"] = "STARTED";
				}
			} else {
				// there are no more letters, this is the end of the final word
				IUInterface::ptr parent_iu = _ib->get(parent_uid);
				parent_iu->payload()["STATE"] = "REALIZED";
			}
			std::cout << " ... done." << std::endl;
		}
	} else {
	}
}

void TextSender::inbuffer_handle_iu_event(IUInterface::ptr iu, IUEventType event_type, bool local)
{
	if (event_type == IU_LINKSUPDATED) {
		std::cout << "links updated" << std::endl;
	} else if (event_type == IU_ADDED) {
		std::string word = iu->payload()["WORD"];
		std::cout << "Received new word: " << word << std::endl;
		publish_text_to_print(word, iu->uid());
	} else if (event_type == IU_RETRACTED) {
		std::string retracted_uid = iu->uid();
	} else {
		std::cout << "(IU event " << iu_event_type_to_str(event_type) << " " << iu->uid() << ")" << std::endl;
	}
}

IUInterface::ptr TextSender::find_last_iu() {
	std::set<IUInterface::ptr> ius = _ob->get_ius();
	for (std::set<IUInterface::ptr>::iterator it = ius.begin(); it!=ius.end(); ++it) {
		if ((*it)->get_links("SUCCESSOR").size() == 0) return *it;
	}
	return IUInterface::ptr();
}

void TextSender::publish_text_to_print(const std::string& text, const std::string& parent_iu_uid) {
	IUInterface::ptr previous_iu = find_last_iu();
	if (previous_iu) {
		// insert a blank if we already have words in the buffer
		IU::ptr iu = IU::create( SEND_CATEGORY );
		iu->payload()["CONTENT"] = " ";
		_ob->add(iu);
		previous_iu->add_link( "SUCCESSOR", iu->uid() );
		iu->add_link( "PREDECESSOR", previous_iu->uid() );
		if (parent_iu_uid != "") iu->add_link( "GRIN", parent_iu_uid );
		previous_iu = iu;
	}
	for (int i=0; i<text.size(); ++i) {
		IU::ptr iu = IU::create( SEND_CATEGORY );
		iu->payload()["CONTENT"] = std::string(1, text.at(i));
		_ob->add(iu);
		if (previous_iu) {
			previous_iu->add_link( "SUCCESSOR", iu->uid() );
			iu->add_link( "PREDECESSOR", previous_iu->uid() );
			if (parent_iu_uid != "") iu->add_link( "GRIN", parent_iu_uid );
		}
		if (previous_iu) std::cout << "previous IU: " << *previous_iu << std::endl;
		previous_iu = iu;
	}
}
	
int main() {
	TextSender sender;
	sleep(1);
	sender.publish_text_to_print("(INIT)");
	std::cout << "Press Ctrl-C to cancel..." << std::endl;
	while (true) sleep(1);
}

int old_main() {
	std::cerr << "TODO: implement Ipaaca C++ test cases." << std::endl;
	return 0;
}

File: ipaacalib/python/test/src/testipaaca.py
# This file is part of IPAACA, the
#  "Incremental Processing Architecture
#   for Artificial Conversational Agents".	
#
# Copyright (c) 2009-2013 Sociable Agents Group
#                         CITEC, Bielefeld University	
#
# http://opensource.cit-ec.de/projects/ipaaca/
# http://purl.org/net/ipaaca
#
# This file may be licensed under the terms of of the
# GNU Lesser General Public License Version 3 (the ``LGPL''),
# or (at your option) any later version.
#
# Software distributed under the License is distributed
# on an ``AS IS'' basis, WITHOUT WARRANTY OF ANY KIND, either
# express or implied. See the LGPL for the specific language
# governing rights and limitations.
#
# You should have received a copy of the LGPL along with this
# program. If not, go to http://www.gnu.org/licenses/lgpl.html
# or write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.	
#
# The development of this software was supported by the
# Excellence Cluster EXC 277 Cognitive Interaction Technology.
# The Excellence Cluster EXC 277 is a grant of the Deutsche
# Forschungsgemeinschaft (DFG) in the context of the German
# Excellence Initiative.

import sys
import time
import unittest

import hamcrest as hc
import ipaaca

def handle_iu_event(iu, event_type, local):
	#print('(IU event '+event_type+' '+str(iu.uid)+')')
	pass

class IpaacaIUStoreTestCase(unittest.TestCase):
	def setUp(self):
		self.ib = ipaaca.InputBuffer('TestIn', ['sensorcategory'])
		self.ib.register_handler(handle_iu_event)
		self.ob = ipaaca.OutputBuffer('TestOut')
		self.sensor_iu = ipaaca.IU('sensorcategory')
		self.sensor_iu.payload = {'data': 'sensordata'}
		time.sleep(0.1)
		self.ob.add(self.sensor_iu)
		time.sleep(0.1)
	def tearDown(self):
		pass
	def testInputBufferContents(self):
		hc.assert_that(self.ib.iu_store, hc.has_key(self.sensor_iu.uid))
		self.assertEqual(len(self.ib.iu_store), 1)
	def testOutputBufferContents(self):
		hc.assert_that(self.ib.iu_store, hc.has_key(self.sensor_iu.uid))
		self.assertEqual(len(self.ob.iu_store), 1)

class IpaacaPayloadTestCase(unittest.TestCase):
	def setUp(self):
		self.ib = ipaaca.InputBuffer('TestIn', ['sensorcategory', 'decisioncategory'])
		self.ob = ipaaca.OutputBuffer('TestOut')
		self.sensor_iu = ipaaca.IU('sensorcategory')
		self.sensor_iu.payload = {'data': 'sensordata'}
		self.ob.add(self.sensor_iu)
		
	def testPayloadContent(self):
		time.sleep(0.1)
		iu_received = self.ib.iu_store.get(self.sensor_iu.uid)
		self.assertEqual(iu_received.payload["data"], 'sensordata')


class IpaacaCommitTestCases(unittest.TestCase):

	def setUp(self):
		self.ib = ipaaca.InputBuffer('TestIn', ['sensorcategory'])
		self.ob = ipaaca.OutputBuffer('TestOut')
		self.iu = ipaaca.IU('sensorcategory')

	def testCommitBeforePublish(self):
		self.iu.commit()
		self.ob.add(self.iu)
		time.sleep(0.1)
		received_iu = self.ib.iu_store[self.iu.uid]
		self.assertTrue(received_iu.committed)

	def testCommitAfterPublish(self):
		self.ob.add(self.iu)
		self.iu.commit()
		time.sleep(0.1)
		received_iu = self.ib.iu_store[self.iu.uid]
		self.assertTrue(received_iu.committed)
	
	def testCommitAndLocalWrite(self):
		self.ob.add(self.iu)
		time.sleep(0.1)
		self.iu.commit()
		try:
			self.iu.payload['data'] = 'updatedData'
			self.fail("Expected an IUCommittedError but it was not raised.")
		except ipaaca.IUCommittedError, e:
			pass
	
	def testCommitAndRemoteWrite(self):
		self.ob.add(self.iu)
		self.iu.commit()
		time.sleep(0.1)
		received_iu = self.ib.iu_store[self.iu.uid]
		try:
			received_iu.payload['data'] = 'updatedData'
			self.fail("Expected an IUCommittedError but it was not raised.")
		except ipaaca.IUCommittedError, e:
			pass


class IpaacaLinksTestCase(unittest.TestCase):
	def setUp(self):
		self.ib = ipaaca.InputBuffer('TestIn', ['sensorcategory', 'decisioncategory'])
		self.ob = ipaaca.OutputBuffer('TestOut')
		self.sensor_iu = ipaaca.IU('sensorcategory')
		self.sensor_iu.payload = {'data': 'sensordata'}
		self.ob.add(self.sensor_iu)
	def tearDown(self):
		pass
	def testSetSingleLink(self):
		time.sleep(0.1)
		self.decision_iu = ipaaca.IU('decisioncategory')
		self.decision_iu.payload = {'data':'decision'}
		self.decision_iu.set_links( { 'grin': [self.sensor_iu.uid] } )
		self.ob.add(self.decision_iu)
		time.sleep(0.1)
		# test received version
		hc.assert_that(self.ib.iu_store, hc.has_key(self.decision_iu.uid))
		received_iu = self.ib.iu_store[self.decision_iu.uid]
		grinlinks = received_iu.get_links('grin')
		hc.assert_that(grinlinks, hc.has_item(self.sensor_iu.uid))
		self.assertEqual(len(grinlinks), 1)
	def testSetAndRemoveSingleLink(self):
		time.sleep(0.1)
		self.decision_iu = ipaaca.IU('decisioncategory')
		self.decision_iu.payload = {'data':'decision'}
		self.decision_iu.set_links( { 'grin': [self.sensor_iu.uid] } )
		self.ob.add(self.decision_iu)
		time.sleep(0.1)
		self.decision_iu.remove_links('grin', [self.sensor_iu.uid])
		time.sleep(0.1)
		# test received version
		hc.assert_that(self.ib.iu_store, hc.has_key(self.decision_iu.uid))
		received_iu = self.ib.iu_store[self.decision_iu.uid]
		grinlinks = received_iu.get_links('grin')
		self.assertEqual(len(grinlinks), 0)


class IpaacaRemoteWriteTestCase(unittest.TestCase):
	def setUp(self):
		self.ib = ipaaca.InputBuffer('TestIn', ['sensorcategory'])
		self.ib.register_handler(handle_iu_event)
		self.ob = ipaaca.OutputBuffer('TestOut')
		self.iu = ipaaca.IU('sensorcategory')
		self.iu.payload = {'data': 'sensordata'}
		time.sleep(0.1)
		self.ob.add(self.iu)
		time.sleep(0.1)
	def tearDown(self):
		pass
	def testRemotePayloadChange(self):
		hc.assert_that(self.ib.iu_store, hc.has_key(self.iu.uid))
		received_iu = self.ib.iu_store[self.iu.uid]
		received_iu.payload['data'] = 'updatedData'
		time.sleep(0.1)
		self.assertEqual(self.iu.payload['data'], 'updatedData')
	def testRemotePayloadReplace(self):
		hc.assert_that(self.ib.iu_store, hc.has_key(self.iu.uid))
		received_iu = self.ib.iu_store[self.iu.uid]
		received_iu.payload = { 'key1': 'value1', 'key2': 'value2' }
		time.sleep(0.1)
		self.assertEqual(len(self.iu.payload), 2)
		self.assertEqual(self.iu.payload['key1'], 'value1')
		self.assertEqual(self.iu.payload['key2'], 'value2')


if __name__ == '__main__':
	unittest.main()

